enhancements/network/on-prem-configurable-lb-dns-stack.md

```diff
@@ -0,0 +1,141 @@
+
+This enhancement suggests a new method for providing the self-hosted stack.
+ 
+*Note:* as per the OCP plan to move API and Ingress to K8S SLB, this direction should be investigated deeper after MetalLb will be integrated.
+
+
+## Motivation
+
+To fully control which customers run our self hosted stack and to be able to serve the needs they really have instead of providing them in a generic way. 
+Additionally, detaching the self hosted stack configuration from the selected platform will allow new customers from UPI (platform: None) to leverage the self-hosted stack.
+
+
+### Goals
+
+- Ability to use the self-hosted stack in all on-prem platforms.
+- Ability to control the self-hosted stack configuration.
+- Retrieve the self-hosted stack status through K8S API.
+- Maintain a single code base/manifests for all the on-prem platforms.
+- Use the self-hosted stack also for UPI deployments.
+
+
+### Non-Goals
+
+## Proposal
+
+The [Keepalived and infra DNS components](https://github.com/openshift/machine-config-operator/tree/master/manifests/on-prem) on the bootstrap node will keep running through the MCO.
+However, the services on masters and workers nodes will be managed by a new operator instead of static pods.
+
+The following functionalities are expected as part of the new operator:
+* Introduce a CRD used by the operator to manage the self-hosted stack.
+* Fields in CR spec that control the self-hosted stack configuration.
+* Fields in CR status that expose self-hosted stack status (e.g: which node holds the API-VIP).
+* Run the self-hosted stack components (Keepalived, Coredns-MDNS, HAProxy, MDNS-publisher) as a K8S applications.
+
+### Kubelet and Operator circular dependency: 
+In order to run the operator's component, Kubelet in each node should first join the cluster by communicating with the control plane, however the components provided by the operator are responsible for enabling that.
+As a result, the below circular dependency should be addressed:
+* Kubelet can't talk to the control plane until the operator has started.
+* Operator can't start until Kubelet can talk to the control plane.
+
+To break the circular dependency, the operator should start running before bootstrap is completed in order to assure that bootstrap owns the API-VIP. In addition, a new /etc/hosts entry should be added to the nodes using MCO ignition.  This /etc/hosts entry is expected to resolve api-int hostname to the API-VIP address. 
+
+### CRD definition
+
+Below is a possible option for a CRD instance of this operator:
+
+
+```yaml
+apiVersion: clusterhosted.operator.io/v1alpha1
+kind: ClusterHostedConfig
+metadata:
+  name: clusterhostedconfig-sample
+spec:
+  dns: Enable
+  loadbalancer: Enable
```

Comment:
This implies we need to hand over the management of the host's `/etc/hosts` and `/etc/resolv.conf` file to the new operator, otherwise we can't update the configuration on the fly. We likely need a new section explaining how we're pointing to the configured LB/DNS.

Reply:
Yep, good point, since the DNS functionality is composed of both host/system config  (resolv.conf for example ) and the operator DNS component, we should cover the E2E design of the DNS.
I'll add a separate section for that purpose.

Reply:
I think what we had discussed previously was that modification of those files would remain in MCO and only the deployment of the pods themselves would be done by the new operator. That way we don't have two separate operators trying to do machine-config things.

To deal with the case where hosted DNS is disabled, we would configure the coredns instance as purely a forwarding server. That way we can unconditionally do the DNS prepending, but if deployers don't want the hosted DNS services they don't have to use them.

Trying to modify the system files conditionally pretty much puts us back in the bad place we were in the first time we tried to do this where we need a bunch of openshift-api changes to provide configuration knobs for this stuff. I would prefer to avoid that if possible.